src/librustc_typeck/check/mod.rs

```diff
@@ -1709,6 +1722,10 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
     }
 
     pub fn expr_ty(&self, ex: &hir::Expr) -> Ty<'tcx> {
+        if let Some(&adjustment::AdjustEmptyToAny(ref t))
+                = self.tables.borrow().adjustments.get(&ex.id) {
+            return t;
+        }
```

Comment:
I suggested this approach. The existing code might misbehave if `expr_ty` is sometimes adjusted.